### PR TITLE
2 Volumes and Volume Rendering

### PR DESCRIPTION
**Pull request by nbaraka**

### CODE REVIEW

**Comment by mathildemerle** [comment]
Fix https://github.com/medInria/medInria-public/issues/603 and i change the title to understand

File: src/layers/legacy/medCoreLegacy/views/interactors/medAbstractLayeredViewInteractor.cpp [hunk 1]
```diff
@@ -37,6 +37,7 @@ medAbstractLayeredViewInteractor::medAbstractLayeredViewInteractor(medAbstractVi
 medAbstractLayeredViewInteractor::~medAbstractLayeredViewInteractor()
 {
     delete d;
+    d =nullptr;
```
Comment (mathildemerle) [inline src/layers/legacy/medCoreLegacy/views/interactors/medAbstractLayeredViewInteractor.cpp:40]:
```suggestion
    d = nullptr;
```

File: src/layers/legacy/medVtkInria/vtkImageView/vtkImageView2D.cxx [hunk 2]
```diff
@@ -2217,12 +2217,12 @@ void vtkImageView2D::RemoveLayer(int layer)
...
-        for ( size_t i(0); i<this->LayerInfoVec.size(); ++i )
-        {
-            if( this->LayerInfoVec[i].Renderer )
-                this->LayerInfoVec[i].Renderer->SetLayer(static_cast<int>(i));
-            this->SetCurrentLayer(static_cast<int>(i));
-        }
+       for ( size_t i(0); i<this->LayerInfoVec.size(); ++i )
+       {
+           if( this->LayerInfoVec[i].Renderer )
+               this->LayerInfoVec[i].Renderer->SetLayer(static_cast<int>(i));
```
Comment (mathildemerle) [inline src/layers/legacy/medVtkInria/vtkImageView/vtkImageView2D.cxx:2223]:
```suggestion
           if( this->LayerInfoVec[i].Renderer )
           {
               this->LayerInfoVec[i].Renderer->SetLayer(static_cast<int>(i));
           }
```

**Comment by Florent2305** [comment]
@nbaraka I'm sorry but I test your branch branch and the issues is still there.
Do you forget a modification into your commit ?

**Comment by nbaraka** [comment]
> @nbaraka I'm sorry but I tested your branch and the issue is still there.
> Did you forget a modification into your PR?

You're right I have commented some line but did not push because this could not be the sol but it only highlight the pb. 
I think I have already explained what seems to be an issue :  interactor are only created for the first view but are deleted each time a data is removed. So if you remove the data to which the interactor is associated then you have the other data that remains without interactor. (

You can try by commented the line 141 (**delete pInteractor;**)  in void medAbstractImageView::removeInteractors(medAbstractData *data)

You should not any pb then. 



I wait your feedback. 
Best regards.

**Comment by nbaraka** [comment]
_> > @nbaraka I'm sorry but I tested your branch and the issue is still there.
> > Did you forget a modification into your PR?
> 
> You're right I have commented some line but did not push because this could not be the sol but it only highlight the pb.
> I think I have already explained what seems to be an issue : interactor are only created for the first view but are deleted each time a data is removed. So if you remove the data to which the interactor is associated then you have the other data that remains without interactor. (
> 
> You can try by commented the line 141 (**delete pInteractor;**) in void medAbstractImageView::removeInteractors(medAbstractData *data)
> 
> You should not any pb then.
> 
This test seems to agree with my hypothese on the bug. 
(T1 and T2 are tow nrrd images) 

Load T1 and then T2 
			select T2 
			close T1 
			no crash 
			but 3D -MSR - black screen
			3D -MIP, ... correct 

			select T1 
			close T1 
			no crash
			but 3D -MSR - black screen
			3D -MIP, ...  correct 

			select T2 
			close T2 
			no crash on 3D -MSR ( correct )
			crash when change to VR

			select T1
			close T2 
			no  crash on 3D -MSR ( correct )
			crash when change to VR


Load T2 and then T1
symmetric 

Best regards.

**Comment by mathildemerle** [comment]
Indeed, the issue https://github.com/medInria/medInria-public/issues/603 is also solved by https://github.com/medInria/medInria-public/pull/625 and https://github.com/medInria/medInria-public/pull/623. Thank you @nbaraka for the tests and investigations :)

**Comment by Florent2305** [comment]
It solve by other PR then I close it

**Comment by nbaraka** [comment]
I think that there is still the pb (i will do the test soon). It migh come from 
       d->primaryInteractorsHash.insert(data, interactor)  in bool medAbstractImageView::initialiseInteractors(medAbstractData *data). 

missing operator==() or not well defined in medAbstractData. 

Best regards.

**Pull request closed without merging**